Feature/for review

iojkgh88 opened the pull request.

iojkgh88 committed:
[#2] 도메인 구성작업

iojkgh88 committed:
[#3] Enum값을 dto로 전달위해 public전환.

iojkgh88 committed:
[#4] Matching 서비스 도메인설계작성
 Member와 생명주기를 겸하는 객체가 많아 Embedded타입의 Entity승격에 의한
 OneToMany 단방향매핑이 다수 진행됨

hongseungjae committed:
feat: 엔티티 생성 및 연관관계 매핑

hongseungjae committed:
feat: 엔티티 수정 및 프로젝트 생성 테스트 코드 작성

hongseungjae committed:
feat: Founder에 projectRole 추가

hongseungjae committed:
feat: Project Builder패턴 적용 및 테스트

hongseungjae committed:
feat: Project Update 기능 추가

hongseungjae committed:
feat: Founder 엔티티와 Participant 엔티티를 Participant로 합침

hongseungjae committed:
feat: Project 서비스계층 추가

hongseungjae committed:
feat: periodDate 타입변경 Long -> LocalDateTime

hongseungjae committed:
feat: Project Controller 추가

hongseungjae committed:
feat: Project Controller 추가

hongseungjae committed:
feat: Project update 로직 수정

hongseungjae committed:
refactor: Project create 로직 수정

hongseungjae committed:
refactor: Project delete 로직 수정

hongseungjae committed:
refactor: Project CreateProjectDto 로직 수정

hongseungjae committed:
feat: ParticipantPosition, ParticipantSkill 추가

hongseungjae committed:
feat: RecruitPosition, RecruitSkillStack 추가

hongseungjae committed:
feat: ProjectController, ParticipantController 작성

hongseungjae committed:
feat: Project find Optional처리 및 NotFoundException 작성

hongseungjae committed:
refactor: ParticipantController 수정

iojkgh88 committed:
[#5] TimePlan 애그리거트 도메인 설계
- Matching -> TimePlan으로 이름변경
- TimePlan은 AbleRoutine을 값타입의 엔티티승격을통한 OneToMany로 진행
- TimePlan과 Event는 1:N 관계로 설정

iojkgh88 committed:
[#5] TimePlan 애그리거트 도메인 테스트
- 객체생성이 용이하지 못하여 Fixture클래스를 만들고 테스트케이스 작성
- TimeBlock, RegularTimeBlock 테스트 완료
- TimePlan 가능시간 생성로직은 아직 미구현되어 테스트통과가 안되고 있음.

iojkgh88 committed:
[#6] TimePlan Controller 및 Service Repository 구현

hongseungjae committed:
feat: 구조변경 및 구성원 CRUD 추가

hongseungjae committed:
refactor: Role 클래스관련 삭제

hongseungjae committed:
feat: Project delete 추가

hongseungjae committed:
feat: Project SkillStack 추가

hongseungjae committed:
feat: Project Recruitment service 추가

hongseungjae committed:
refactor: NonNull -> NotNull 수정

hongseungjae committed:
refactor: Post 요청 시 Location 헤더에 URI 추가

hongseungjae committed:
refactor: Exception 코드 정리

hongseungjae committed:
refactor: Application Exception 제외 코드 정리

hongseungjae committed:
refactor: Exception 코드 정리

hongseungjae committed:
refactor: Project와 Participant로 패키지 분리, GlobalSetting 패키지 추가

hongseungjae committed:
feat: Project 좋아요 기능 추가

iojkgh88 committed:
[#7] 테스트를 위한 member_prototype 구현

iojkgh88 committed:
[#9] timeplan에서 extractAbleTimeBlock 구현완료

hongseungjae committed:
refactor: Project Service계층에서 presentation계층 의존성 제거(service dto생성)

hongseungjae committed:
refactor: Participant Service계층에서 presentation계층 의존성 제거(service dto생성)

hongseungjae committed:
refactor: orElseThrow 활용

hongseungjae committed:
test: ProjectServiceTest, ParticipantServiceTest 테스트 코드 작성

hongseungjae committed:
test: SpringDataJpa 사용

iojkgh88 committed:
[#10] timeplan/ schedular 구현 / test실패

hongseungjae committed:
Merge remote-tracking branch 'origin/feature/forReview' into feature/project4

# Conflicts:
#	build.gradle
#	src/main/java/com/flab/quicktogether/matching/domain/Role.java
#	src/main/java/com/flab/quicktogether/member/domain/Member.java

hongseungjae committed:
test: TimePlan과 Project Merge

hongseungjae committed:
Merge pull request #2 from f-lab-edu/feature/project4

Project Merge

iojkgh88 committed:
[#11] event 연관관계 오타 수정

iojkgh88 committed:
Merge branch 'feature/forReview' of github.com:f-lab-edu/quick-together into feature/forReview

iojkgh88 committed:
test에 lombok의존성 추가

iojkgh88 committed:
로그인체크 인터셉터 충돌로 webConfig에서 /timeplan path에서만 작동되도록 임시수정

f-lab-fi commented on src/main/java/com/flab/quicktogether/common/TimeSection.java:
```diff
@@ -0,0 +1,30 @@
+    private LocalTime startTime;
+    private LocalTime endTime;
+
+    // 유효성 검사 startTime.isBefore(endTime)
+    public TimeSection(LocalTime startTime, LocalTime endTime) {
+        if (startTime.isBefore(endTime)) throw new IllegalTimeDateException();
+        this.startTime = startTime;
+        this.endTime = endTime;
+    }
+
+
+    // 가능한 조건 |가능한시간 --- 이벤트시작시간 ---- 이벤트 종료시간 --- 가능한시간
+    public boolean isAvailableWith(TimeSection eventTimeSection) {
+        if (this.startTime.isAfter(eventTimeSection.startTime) && this.endTime.isBefore(eventTimeSection.endTime)) {
+            return true;
+        }
+        return false;
+    }
```
두 메서드 중 위에있는 메서드에 대한 유효성 검사는 시작 시간과 종료 시간이라는 두 시간 중 시작 시간이 더 과거 시간이어야 한다는 내용은 일반적으로 통용 될 수 있는 개념이라서 'common' 이라는 위치에도 적절해보입니다.
그러나 아래쪽에 있는 메서드는 '이벤트 시간은 가능한 시간의 안에 포함되어야 한다'라는 도메인 지식에 해당한다고 생각해야할 것 같습니다. 그럼 여기있는게 적절하지 않겠죠.

사실 첫번째 것도 생각해보면 startTime과 endTime 사이의 관계 역시 둘 모두 그냥 '시간'이라는 개념이기 때문에 시작 시간이 종료 시간보다 나중이라는 것도 도메인 지식이라고 생각할 여지가 있습니다. :)

iojkgh88 replied:
넵 현재는 TimeBlock으로 바뀌었는데, common에서 TimePlan의 도메인으로 옮겼습니다.
merge하는 과정에서 기존것이 남아있고 사라지지 않았네요. 

활용하기에는 결국 Meeting시간을 제안해주기위한 기반자료로서 활용되기 때문에 common보다는 도메인쪽의 위치가 적절하다고 판단했습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/globalsetting/exception/ExceptionResponse.java:
```diff
@@ -0,0 +1,52 @@
+
+@Data
+@NoArgsConstructor
+public class ExceptionResponse {
+    private LocalDateTime timestamp = LocalDateTime.now();
+
+    private int status;
+    private String error = "";
+    private String code = "";
+    private String message = "";
+    private String path = "";
```
ExceptionResponse는 Body로 사용되는 DTO인데, 여기에 status가 들어갈 필요가 있을까요?
HTTP 상태 코드는 응답 헤더에도 들어갈텐데, 바디에도 있다면 클라이언트 쪽에서 바디에 있는 status를 사용하게될 우려가 있습니다.

hongseungjae replied:
아 넵
응답 헤더에 어차피 status가 들어있어서 반환 해줘야 하는지 고민 됐던 부분인데
오히려 두 개 다 들어있을 경우 클라이언트에서 무엇을 사용 해야 할 지, 혹은 두 개가 다른 의미인지 헷갈릴 포인트 같습니다.

f-lab-fi replied:
넵. 가능하면 바디에는 넣지 않는게 좋습니다.
HTTP 상태 코드는 오직 헤더에 있을때 좋은 상황을 많이 만듭니다.
클라이언트쪽 코드에서도 HTTP 상태 코드를 통해 요청이 성공인지 실패인지를 핸들링 할 수 있는 매커니즘을 가지고 있습니다.
아래 링크를 참고해보세요.

https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/error_event

f-lab-fi commented on src/main/java/com/flab/quicktogether/globalsetting/exception/GlobalResponseEntityExceptionHandler.java:
```diff
@@ -0,0 +1,92 @@
+                                                        HttpHeaders headers,
+                                                        HttpStatusCode status,
+                                                        WebRequest request) {
+        log.warn(ex.fillInStackTrace().getMessage());
+        String message = "'"+ex.getValue()+"' to required type '"+ex.getRequiredType()+"'";
+
+        if (ex instanceof MethodArgumentTypeMismatchException) {
+            MethodArgumentTypeMismatchException castEx  = (MethodArgumentTypeMismatchException) ex;
+            message = "'"+ex.getValue()+"' to required type '"+castEx.getName()+"'";
+        }
```
문자열 조합하실땐 가독성을 위해서라도 위에 사용하신 것 처럼 String.format를 사용하시는게 좋을 것 같습니다. :)

hongseungjae replied:
넵 수정하도록 하겠습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/matching/domain/SkillStack.java:
```diff
@@ -0,0 +1,7 @@
+package com.flab.quicktogether.matching.domain;
+
+public enum SkillStack {
+        JAVA, KOTLIN, SPRING, JPA,
+        JAVA_SCRIPT, NODE_JS, EXPRESS_JS, NEST_JS, REACT_JS, VUE_JS, SVELTE_JS, ANGULAR_JS,
+        PYTHON, DJANGO, FLASK, TENSOR_FLOW
+}
```
https://github.com/f-lab-edu/quick-together/pull/3/files#diff-405043fff8225810476862db91061cf2324b5656ebe02daa572f2a760a635891

위 클래스랑 서로 다른 클래스인가요~?

iojkgh88 replied:
현재 matching 패키지가 존재하지 않습니다. 삭제하도록 하겠습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/meeting/domain/Meeting.java:
```diff
@@ -0,0 +1,25 @@
+import java.time.LocalDate;
+
+
+public class Meeting {
+    Long id;
+    Long meetingNo; // 클라이언트가 볼수있는 미팅고유번호
+    String meetingName; // 미팅이름
+    LocalDate meetingDate; // 미팅시작일
+    TimeSection meetingTime; // 미팅시간
+
+    String description; // 상세설명
+
+    CUDTimeStampInfo meetingTimeStampInfo; // 생성, 수정, 삭제에 대한 TimeStamp 정보
+
+    //    private AlarmSettingInfo 추후 알람기능 생성시
+}
+
+enum MeetingStatus {
+    APPROVED, DENIED, VOTING, STARTED, CLOSED
+}
```
이 클래스는 아직 사용하기 전인거죠~?

iojkgh88 replied:
아아 넵 맞습니다. 
미완성된부분은 주석으로라도 표시해드렸어야 보기 편하셨을텐데 
미처 생각이 닿지 못했네요.

앞으로는 표시해두겠습니다! 

Meeting과 Scheduling은 따로 체크 안해주셔도 될것같습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/member/presentation/MemberController.java:
```diff
@@ -0,0 +1,26 @@
+
+    private final MemberService memberService;
+
+    @NotRequiredLoginCheck
+    @RequestMapping(path = "/login", method = RequestMethod.GET)
+    public ResponseEntity login(HttpSession httpSession) {
```
나중에 로그인 처리할 때에도 GET 메서드로 유지될 수 있을까요?
유지될 수 없다면 왜 유지될 수 없는지에 대해서도 생각해보시면 좋을 것 같습니다.

iojkgh88 replied:
member객체로 다른 서비스에서 /login만으로 로그인된것처럼 간이로 구현된것이라 get이라고 했는데
나중이 아니더라도 HttpSession에 저장되는 이상 POST로 보는것이 맞네요.
수정하겠습니다.

f-lab-fi replied:
그것도 그렇고, 로그인할 때는 인증 정보(비밀번호 같은 것)를 보내게 됩니다.
그때 인증 정보를 GET으로 보내는건 인증 정보를 URL에 노출시키는 문제가 있습니다.

iojkgh88 replied:
앗 넵 해당 익히 공부한 내용인데 솔직하게 코드작성시에는 생각을 못했던것 같습니다.
정확히 인지했습니다!

f-lab-fi commented on src/main/java/com/flab/quicktogether/participant/application/ParticipantService.java:
```diff
@@ -0,0 +1,173 @@
+        participantRepository.save(Participant.addMember(project, member));
+    }
+
+    private void checkProjectParticipation(Long projectId, Long memberId) {
+        Optional<Participant> participant = participantRepository.findByProjectIdAndMemberId(projectId, memberId);
+        if (participant.isPresent()) {
+            throw new DuplicateProjectParticipationException(DUPLICATE_PROJECT_PARTICIPATION);
+        }
+    }
```
`if (participant.isPresent())`는 전형적인 Optional 안티패턴입니다.
어떻게 작성하는게 좋을까요~?

hongseungjae replied:
프로젝트에 중복참여를 방지하기 위한 메서드로
if문으로 participant에 값이 들어있는지 확인하는것보다
ifPresent를 활용하는 것이 더 좋을 것 같습니다
```
        participantRepository.findByProjectIdAndMemberId(projectId, memberId).ifPresent(joinedParticipant -> {
                    throw new DuplicateProjectParticipationException(DUPLICATE_PROJECT_PARTICIPATION);
                });
```

f-lab-fi commented on src/main/java/com/flab/quicktogether/participant/application/ParticipantService.java:
```diff
@@ -0,0 +1,173 @@
+
+    }
+
+    private Participant findParticipant(Long projectId, Long memberId) {
+        findProject(projectId);
+        findMember(memberId);
+        return participantRepository.findByProjectIdAndMemberId(projectId, memberId).orElseThrow(
+                () -> new ParticipantNotFoundException(PARTICIPANT_NOT_FOUND));
+    }
+
+    private Project findProject(Long projectId) {
+        return projectRepository.findById(projectId).orElseThrow(
+                () -> new ProjectNotFoundException(PROJECT_NOT_FOUND));
+    }
+
+    private Member findMember(Long memberId) {
+        return memberRepository.findById(memberId).orElseThrow(
+                () -> new MemberNotFoundException(MEMBER_NOT_FOUND));
+    }
```
공통적으로 활용되는 로직을 잘 묶어보셨네요.
간단한 코드지만, Project가 없을 때는 Project가 없다는 예외, Member가 없을때는 Member가 없다는 예외
그리고 해당 멤버가 해당 프로젝트에 참여중인게 아니라면 참여중이지 않다는 예외가 간단한 코드지만 분명하게 전달되고 있네요. :)

f-lab-fi commented on src/main/java/com/flab/quicktogether/participant/domain/Participant.java:
```diff
@@ -0,0 +1,78 @@
+        this.participantRole = participantRole;
+    }
+
+    public static Participant addMember(Project project, Member member){
+        return new Participant(member, project, ParticipantRole.ROLE_USER);
+    }
```
Participant는 그 의미가 Project에 참여한 Member를 의미하는데, 그렇다면 addMember라는 이름은 조금 어색한 느낌이 있습니다.
어떤 이름이 더 적합할지는 저도 고민되네요. 이부분은 같이 고민해보면 좋을 것 같습니다.

iojkgh88 replied:
승재님과 이야기 나누어본 결과 
Participant의 권한등급이 Enum으로 나뉘어지고있어 
ADMIN이 참가요청에 대한 수락의 의미로 addMember메소드를 활용하실 예정입니다.

따라서 permitParticipant로 메서드명을 수정하는것으로 결정해보았습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/participant/presentation/dto/request/ChangeParticipantRoleRequest.java:
```diff
@@ -0,0 +1,18 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class ChangeParticipantRoleRequest {
+    @NotNull
+    private ParticipantRole participantRole;
+
+    public ChangeParticipantRoleRequestDto toServiceDto() {
+        return new ChangeParticipantRoleRequestDto(this.getParticipantRole());
+    }
+}
```
그러고보니 컨트롤러에서 사용하는 DTO와 서비스에서 사용하는 DTO를 나누신건가요?
너무 DTO가 많아진다는 느낌이 들진 않으신지 두분 생각이 궁금합니다.

hongseungjae replied:
아 넵
컨트롤러에서 서비스로 데이터를 전달할 때 서비스층 dto로 변환 후 전달합니다.

먼저 이렇게 한 첫번째 이유는
presentation -> application와 같은 흐름에서 application계층에서 presentation계층의 dto를 알고 있다는게 흐름에 맞지 않는다고 생각했습니다.
하지만 이 부분은 dto를 어디 계층까지 허용할 것인가에 대한 문제로 크게 상관은 없을 것 같습니다.

두번째 이유는 초기 개발할 시 앞단의 컨트롤러에서 매핑 url이 바뀌거나, dto가 바뀔 시 
전달하는 인자값이 조금씩 달라져서 서비스계층이 사용하는 데이터는 정해져 있음에도 계속적으로 서비스 계층에서 수정이 필요했습니다.
그래서 서비스계층으로 dto를 넘기기전에 서비스계층의 dto로 바꿔서 전달하는 방법을 사용했습니다.

그런데 어느정도 컨트롤러에 들어오는 데이터 포맷이 정해지고 나서는 굳이 이렇게 넘겨야될까 라고 생각됩니다.
거의 같은 코드를 두 번씩 작성해야 되는 불편함이 있는 것 같습니다.

iojkgh88 replied:
공통의 전제는 해당 RequestDTO는 캡슐화를 포함하여, Service레이어에서 Repository의 조회를 수반함으로 계층구조를 준수하려면 DTO를 Service객체의 메소드인자로 넘겨주어야 한다는 점입니다.

승재님 말씀대로 DTO를 전달하는 경우 문제점은 DTO의 변경이 Service객체에 영향을 미쳐서 OCP를 위반하기 때문입니다.
승재님께서 Service레이어에서 사용되는 DTO를 두시는것은 ServiceDTO는 일종의 Adapter패턴에서의 어댑터로 생각됩니다. 

그럼에도 저의 경우는 DTO를 그대로 Service객체로 반환하는 전략을 택했습니다. 
이유는 아래와 같습니다.

1. DTO의 변화가 Service객체의 변화에 영향을 주지 않아야 하는점인데 컨트롤러단 DTO에 변환메서드를 넣는것 만으로도 OCP는 충족될 수 있습니다.
2. 아래와 같이 곰곰이 생각해보았는데  DTO의 변화는 거의 대부분 서비스의 변경을 수반할수밖에 없지 않을까 생각됩니다.
 case 1. DTO의 상태값이 수정이 Entity 내부에 이미 있는 정보 안에서 일어나는 경우
- DTO의 상태값이 적어졌다면 도메인 내부에서 해당 기능이 더이상 필요가 없어졌거나 다른외부시스템을 통해 가져오는 경우이기에 서비스단의 수정이 필요합니다. 
- DTO의 상태값이 늘어났다면 Service객체에서 DTO를 Entity로 변환하고 기존에 어떤로직에 의해 존재하는 상태값 할당과정을 무시해야할텐데 변환된 Entity가 상태값을 기존 서비스 로직에 따라 할당받으면 결국 DTO의 수정이 의미가 없어집니다. 

 case 2. DTO의 상태값이 Entity의 변경에 의해서 일어나는 경우
- Controller단 DTO내부의 Entity변환 메서드를 만드는 것 만으로도 OCP는 지켜질것이 일부 존재할수 있을것으로 보입니다.
- 항상 변환메서드나 Adapter만으로 Service로직의 변화를 막을 수는 없을것으로 보입니다. 

그리고 Service DTO가 필요한 경우도 생각해보았는데
Controller DTO가 너무 비대해져서 서비스의 분기가 필요하다면, 별도의 serviceDTO를 만드는것이 의미있어보입니다.
엄밀히 해당 DTO를 어디서 처리하는지 분배하는것도 일종의 Service의 역할로 보이긴 합니다.
다만, 규모가 크고 담당자가 다른경우에 일정에 제한이 있다면 컨트롤러단에서 DTO를 따로 나누어서 보내주는 형태로 처리하게될 것 같습니다.

f-lab-fi replied:
넵 좋습니다.
두분 말씀하신 내용 모두 고려해봐야하는 내용입니다.
결국 이런 문제는 트레이드 오프가 필요합니다. 객체지향적인 가치를 중요시해야하는 상황이라면 두분 말씀하신 것과 같은 논리로 나눠볼 수 있는거고, 비슷한 코드의 수가 너무 많아져 관리가 어려워진다면 반대로 합칠 수도 있습니다.
답이 정해져 있는 문제는 아니니 두분이 생각하는 논리가 있다면 잘 작성하신겁니다. :)

f-lab-fi commented on src/main/java/com/flab/quicktogether/timeplan/application/IteralScheduler.java:
```diff
@@ -0,0 +1,58 @@
+        return getIncludeEventTime(candidateTimes, ableTimeblocks);
+    }
+
+    private List<TimeBlock> getCandidateTimes(MinuteUnit minuteUnit, LocalDate targetStartDate, LocalDate targetEndDate, Integer durationMinutes) {
+
+        int unitValue = minuteUnit.getUnitValue();
+        LocalDateTime target = LocalDateTime.of(targetStartDate, LocalTime.MIN);
+
+        LocalDateTime endTarget = LocalDateTime.of(targetEndDate.plusDays(1L),LocalTime.MIN)
+                .minusHours(getHour(durationMinutes))
+                .minusMinutes(getMinute(durationMinutes));
+
+        List<TimeBlock> candidateTimes = new ArrayList<>();
+        while (target.compareTo(endTarget) <= 0) {
+            candidateTimes.add(new TimeBlock(target, target.plusMinutes(durationMinutes)));
+            target = target.plusMinutes(unitValue); // TODO LocalDateTime이 불변객체라 Iterate시 GC부담이 계속될듯 해결필요
+        }
+        return candidateTimes;
+    }
```
[저번에 질문 주셨던 것](https://f-lab-edu.slack.com/archives/C03VAUUD8HJ/p1671442843464359)과 비슷한 맥락인 것 같네요.
실제 GC 부담이 되는지는 실제로 테스트를 해봐야 확실해 지는 부분입니다.
그래도 한번 이런 상황을 해결할 수 있는 관점에 대해 이야기 해볼게요.

이 로직이 지금 제가 이해한 바로는 아래와 같은 로직 같은데 맞을까요?

'시작일시와 종료일시 사이에 있는 시간 중 단위시간(unitValue)에 해당하는 시간들을 candidateTimes에 저장한 후 반환한다.'
라는 개념인 것 같은데요, 이거 어차피 클라이언트에게 되돌려주는 연산 아닌가요?
그렇다면.. 이 로직 자체를 서버가 아니라 클라이언트에게 전가하는 방법도 고려해볼 수 있습니다.
어떻게 클라이언트에게 전가할 수 있을까요?

(제가 잘못 이해하고 이야기하는 것일 수도 있습니다. 혹시 잘못된 내용 있으면 이야기주세요~)

iojkgh88 replied:
클라이언트에게 돌려줄 연산은 공통된 가능시간만 전달하고자 의도했고 CandidateTime은 이를 위한 중간연산값에 해당합니다. 

스케줄링은 미팅생성을 위해 고안되고 있습니다. 
클라이언트 측에서 "회의시간은 대략 1시간정도 걸릴건데 대충 다음주중이면 될것같아." 라고 요청한다면, 각 회의참여자들간의 가능시간정보를 비교해서 공통적으로 가능한 시간대를 제안해서 보여준다는것이 핵심 비즈니스 로직입니다.

이때 사용할 전략으로 아래와 같이 선택했습니다. 
1. 다음주 중이라는 기간내에 설정된 단위시간만큼의 모든 CandidateTime을 생성합니다. 이에 따라 공통으로 가능한시간이 반영되지 않은 모든 단위시간을 생성합니다.
2. CandidateTime마다 모든 구성원의 가능시간 안에 모두 들어와있는지 체크후 모두 들어와있다면 최종 가능시간을 도출합니다.

네이밍을 좀더 명확하게 하는것이 쉽지 않네요. 좀더 다듬어보겠습니다!

f-lab-fi commented on src/main/java/com/flab/quicktogether/timeplan/domain/Event.java:
```diff
@@ -0,0 +1,84 @@
+        }
+    }
+
+    private String nullProcess(String eventName) {
+        if (eventName == null) {
+            eventName = "CONFIDENTIAL";
+        }
+        return eventName;
+    }
```
eventName이 null인 경우 회의 이름이 비공개인 건가요?
어떤 의도인지 궁금합니다~

iojkgh88 replied:
event는 회의뿐만이 아니라 외부 캘린더 연동까지 고려했을때 포함될수있는 모든 일정에 대한 도메인으로 생각하고 있습니다.
가능한 시간을 추출해 내기 위해서 AbleRoutine상에있는 정기 가능시간에서 event를 모두 제외해주어야 실질적인 가능시간이 도출되기때문입니다. 

일정이 있지만 클라이언트 측에서 이름을 입력하길 원하지 않는다면 
클라이언트 측에서 비공개를 의도한것이 아닐까 해서 작성해보았습니다. 

그런데 이름입력을 안했는데 CONFIDENTIAL이라고 쓰는게 조금 무거운 의미로 다가오네요.
좀더 네이밍을 가다듬어볼 필요가 있을것 같다고 생각이 듭니다.

f-lab-fi replied:
그럼 이름이 비공개라는 의미가 되어야할텐데, 클라이언트 코드쪽에

eventName === "CONFIDENTIAL" 이면 비공개 회의를 의미하는 UI 표현

코드가 들어가야할 것 같습니다.
차라리 eventName 자체를 null로 처리하거나 빈 문자열로 보내주는게 명시적으로 비어있다는 의미를 줄 수 있을 것 같아요.
혹시 클라이언트가 실제로 이벤트 이름을 CONFIDENTIAL로 지정하는 상황이라면 어떨까요 ㅎㅎ

iojkgh88 replied:
파이님 말씀을 생각해보니 제가 클라이언트의 의도를 마음대로 추측하지 않았나 생각이 됩니다.
의도가 명시적으로 비공개라면 비공개가 맞지만 적지 않아도 되는데 null포인트를 피하려면 공백이 가장 최적으로 보입니다.

f-lab-fi replied:
아니면 명시적으로 공개되지 않은 이름이라는 것을 의미하는 필드를 추가해도됩니다~

f-lab-fi commented on src/main/java/com/flab/quicktogether/timeplan/domain/Event.java:
```diff
@@ -0,0 +1,84 @@
+        }
+    }
+
+    @Override
+    public int compareTo(Event o) {
+        return this.timeBlock.compareTo(o.getTimeBlock());
+    }
```
Comparable 잘 구현해주셨습니다~
equals와 마찬가지로 getter를 노출하지 않고도 비교할 수 있겠죠.

iojkgh88 replied:
가능시간을 연산하는 과정이다보니 순서대로 정렬하기 위해서는 구현이 필연적이었던것 같습니다.
자바의 신 컬렉션 공부할때 Comparable구현이 필요한 것을 이해했는데 진행하면서
Comparable과 Comparator 간에 어떤것을 선택할지 고민했습니다.

Comparator는 객체다보니 정렬로직등을 상황에따라 전략을 유연하게 바꾸어줄수 있긴 하지만, 클래스가 많아질수있고
핵심적인 정렬비교는 하나고 참조자료형만 사용하기 때문에 클래스 내에 구현하는 Comparable을 사용했습니다.  

공부가 없었다면 저도 값을 하나하나 불러오는식으로 고민했을것 같습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/timeplan/domain/TimePlan.java:
```diff
@@ -0,0 +1,130 @@
+     */
+    private void verifySeperated(List<AbleRoutine> ableRoutines) {
+        Collections.sort(ableRoutines);
+        for (int i = 0; i < ableRoutines.size()-1; i++) {
+            RegularTimeBlock cur = ableRoutines.get(i).getRegularTimeBlock();
+            RegularTimeBlock next = ableRoutines.get(i + 1).getRegularTimeBlock();
+
+            if (!cur.isSeperatedFrom(next)) {
+                throw new AbleRoutineOverlapException();
+            }
+        }
```
이 코드를 좀 개선해보려면 어떻게할 수 있을까요?
분명 작성하시면서 마음에 안드셨지만, 어쩔 수 없이 앞뒤를 비교해야하다보니 이렇게 작성하셨을 것 같아서 여쭤봅니다. :)

iojkgh88 replied:
제 마음을 정확히 읽으셨네요.. ㅠ
Stream을 이용해서 전후비교할 방법이 없을까 생각해보았는데 전후비교를 위한 로직이 없어서
reduce로 우회할수 없나 생각해봤습니다. 
그런데 BinaryOperator에 return을 두번째 값으로 하고 사이에 초기값과 비교해서 exception을 던지는 식으로 해보았는데 Exception이 아니라 ExceptionInInitializerError가 발생했습니다. 

클래스내용 확인해보니 정적변수의 초기화나 정적변수 평가시에 예외가 발생하는경우에 던져지는 에러입니다. 
reduce가 내부적으로 static초기화를 사용하는것인지 원인을 찾기 어렵습니다.  
ReduceOps라는 내부 정적클래스에서 문제가 있는것으로 추측되는데 아직 원인파악이 안되어서 오류재현은 아직못해보았습니다.

당장 리팩토링한다면 iterator()를 사용해서 while문 이용이 적절해보입니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/timeplan/domain/value_type/TimeBlock.java:
```diff
@@ -0,0 +1,201 @@
+        this.endDateTime = endDateTime;
+    }
+
+    private void verifyNaturalTimeOrder(LocalDateTime startDateTime, LocalDateTime endDateTime) {
+        if (startDateTime.isAfter(endDateTime) || startDateTime.equals(endDateTime)) {
+            throw new NotNaturalTimeOrderException();
+        }
+    }
```
아마 서로 작성하신 분이 달라서 이름이 달라진 것 같은데, 유효성 검사 관련 코드들 메서드 이름은 통일해보시면 좋을 것 같습니다~
https://github.com/f-lab-edu/quick-together/pull/3/files#diff-17521c2ca7338b94f71f68182710927eb44a676c4f6b13bb66f9940787152b22R59

valid? validate? verify? 등

iojkgh88 replied:
기존작성은 valid로 했었는데 DDD 시작하기 책에 verify로 유효성검사하는 예시가 나와있어서 추후에 verify로 작성이 되었습니다.
의도적으로 다르게 쓴게 아니다보니 코드작성자가 아닌분들이 보았을때는 valid와 verify가 무슨차이인지 불필요한 생각이 들겠네요.
통일해보도록 하겠습니다!

f-lab-fi commented on src/main/java/com/flab/quicktogether/timeplan/domain/value_type/TimeBlock.java:
```diff
@@ -0,0 +1,201 @@
+
+        ArrayList<TimeBlock> trimmedBlocks = new ArrayList<>();
+
+        if (this.equals(target) || this.isIncludeIn(target)) {
+        } else if (this.isOverlappedWithStartEqualBy(target) || this.isOverlappedForwardBy(target)) {
+            trimmedBlocks.add(this.remainBackward(target));
+
```
여기서

```
if (this.equals(target) || this.isIncludeIn(target)) {
    // do nothing
}
```

처럼 아무것도 안한다는걸 명시적으로 적어주면 좋을 것 같습니다.
순간 중괄호가 잘못된건가 헷갈렸어요.

iojkgh88 replied:
아아 넵 주석을 이렇게 활용하면 좋군요

f-lab-fi commented on src/main/java/com/flab/quicktogether/timeplan/domain/value_type/TimeBlockOperator.java:
```diff
@@ -0,0 +1,145 @@
+//        this.timeBlocks.add(otherTrimmedBlock);
+//    }
+
+    @Getter @Setter
+    static class OperableTimeBlock implements Comparable<OperableTimeBlock> {
+        private LocalDateTime startDateTime;
+        private LocalDateTime endDateTime;
+
+        private OperableTimeBlock(LocalDateTime startDateTime, LocalDateTime endDateTime) {
+            this.startDateTime = startDateTime;
+            this.endDateTime = endDateTime;
+        }
+
+        public static OperableTimeBlock convertFrom(TimeBlock timeBlock) {
+            LocalDateTime startDateTime = timeBlock.getStartDateTime();
+            LocalDateTime endDateTime = timeBlock.getEndDateTime();
+
+            return new OperableTimeBlock(startDateTime, endDateTime);
+        }
+
```
이 내부 클래스가 가지고 있는 코드들은 TimeBlock 클래스와 거의 동일한 것 같은데 어떤 구조인거죵..?

iojkgh88 replied:
TimeBlock 연산과정에서 불변객체를 연산해서 계속 객체가 생성되는것이 부담되어서
가변객체로 변환해서 StringBuilder처럼 연산할수있게 해볼까 만들다가 만 클래스입니다. 

현재는 연산에 활용되지 않은 프로토타입으로 보시면 될것같습니다.

hongseungjae committed:
refactor: isPresent -> ifPresent 활용

f-lab-fi commented on src/test/java/com/flab/quicktogether/timeplan/domain/TimePlanTest.java:
```diff
@@ -0,0 +1,54 @@
+
+    }
+
+    //만기일에 expiredException이 존재해서 항상 성공하는 테스트를 만들려면 상대적인 시간으로 생각해야하는데 테스트짜기 어렵다.
+    @Test
+    @DisplayName("절대타임블록으로 변환하면 가능시간에서 이벤트를 제외한 시간을 반환한다.")
+    public void extractAbleTimeBlock() throws Exception{
+        //given
+        Member testMember = TEST_MEMBER;
+        List<AbleRoutine> dayEqualRoutines = DAY_EQUAL_ROUTINES;
+        List<Event> plannedEvents = PLANNED_EVENTS;
+
+        TimePlan timePlan = new TimePlan(TEST_MEMBER, DAY_EQUAL_ROUTINES);
+        List<TimeBlock> expectedAbleTimeBlock = EXPECTED_ABLE_TIME_BLOCK;
+
+
+        LocalDate fixedDate = FIXED_DATE;
+        //when
+        List<TimeBlock> result = timePlan.extractAbleTimeBlock(PLANNED_EVENTS, FIXED_DATE, FIXED_DATE);
+
+        //then
+        Assertions.assertThat(result).containsExactly(expectedAbleTimeBlock.toArray(new TimeBlock[0]));
+    }
```
이 테스트 테스트 돌아가는 시간 기준으로 바꿔보실건가요~?

iojkgh88 replied:
ExpiredTimeException을 피하기 위해서 적용된 시간은 항상 내일이 되도록 Fixture클래스에서 동적으로 생성하도록 했습니다. 
```java
public static final LocalDate FIXED_DATE = LocalDate.now().plusDays(1L);
```

이럴경우 해당 Fixture를 상수로 놓을수 있는지 컨벤션에 대한 고민이 있었습니다. 
객체생성이 단발적으로 끝나기 때문에 테스트기준으로 볼때는 상수라 판단하고 진행하였습니다. 
https://github.com/f-lab-edu/quick-together/pull/3/commits/0da4f0195e74a13f948ca0f093c169d7a1c8e2d1#diff-4d13c7c139f06b8e3f90e69634fa2f1ec7f728cd29ec1a3b1f233be2804a5d38

f-lab-fi commented:
@gggorock @hongseungjae 

개발하시느라 고생 많으셨습니다.
코드 작성하신 것 관련해서 총평을 드리자면,

1. 전반적으로 제가 단축 URL 프로젝트 하면서 말씀드렸던 내용들이 적절히 반영되어 있어서 좋았습니다.
2. 정상적으로 코드가 동작하는 상황, 예외적인 상황에 대해 크게 빠지는 내용 없이 꼼꼼하게 테스트 코드를 작성하신 것 같다고 생각됐습니다. 물론 이게 모든 엣지 케이스에 대해 테스트가 되었다는 의미는 아닙니다. 그래도 주요 기능들에 대해 테스트는 잘 이루어지고 있다고 생각합니다.
3. 위에 리뷰에도 언급드린 내용이긴 한데, 컨트롤러에 대한 DTO와 서비스에 대한 DTO가 모두 있어서 코드 숫자가 과도하게 많아진 것 같은 느낌이 들긴 합니다. 이건 실제 개발하는 사람이 체감해야하는 문제이니 두분 의견이 더 중요한데요, 나중에 코드의 수가 너무 많아 관리가 어려워진다면 합치는걸 고려해보시기 바랍니다.
4. 예외 처리에 대한 핸들러 코드를 못본 것 같은데, 예외 처리에 대한 핸들러 코드도 작성해보시면 좋을 것 같습니다.
5. 전반적인 코드의 네이밍, 일관성이 지켜져서 읽기 좋은 코드였습니다.
6. 테스트 코드 관련해서는 아마 승재님이 작성하셨던 코드 였던 것 같은데, DisplayName을 세원님이 작성하신 것처럼 바꿔보시면 어떨까 싶습니다. 서로 테스트 코드 작성하신 것 한번씩 읽어보시면 좋을 것 같은데요, 승재님이 작성하셨던 테스트 코드의 DisplayName들을 보면 예외가 발생하는 상황은 무엇을 테스트 하려는지 이름만으로도 명확하게 느껴지지만 성공하는 상황에 대한 테스트는 이름만으로는 그 의도를 파악하기 어려웠습니다.

리뷰 드릴만한 내용은 이정도고, 궁금한 내용 있으면 또 질문 주세요. :)

iojkgh88 committed:
fix : 불필요한 파일 삭제

iojkgh88 committed:
refactor : LoginCheckArgumentResolver 네이밍 변경

이름착오로 멤버아이디를 불러올수있는 이름으로 리팩토링

iojkgh88 committed:
fix : 로그인기능 메서드 POST로 변경

간이로그인이라 하더라도 session에 값을 등록해주기 때문에 GET이 적절하지 않아 변경

iojkgh88 committed:
fix : 스케줄링기능 while조건변경

while문 조건착오로 인한 테스트 실패로 변경

iojkgh88 committed:
refactor : TimePlan service메소드 이름개선

DB와 같은 네이밍이 아닌 비즈니스로직에 적합한 형태로 변경

iojkgh88 committed:
test : TimeBlock 리스트 안 포함여부 테스트 추가

스케줄링서비스를 만드는 과정에서 진행

iojkgh88 committed:
test : TimePlan Repository Test

테스트 추가
"timePlan 을 저장하고 찾으면 기존 저장된 timePlan을 반환한다."

iojkgh88 committed:
style : 아무것도 하지않음을 뜻하는 주석 추가

아무것도 적혀있지 않음이 의도된것임을 명시적으로 표현하기 위하여 주석추가

iojkgh88 committed:
style : 프로토타입이라 개발중임을 주석으로 표시

추후 리뷰어의 리뷰확인이 필요여부를 주석으로 명시

iojkgh88 committed:
refactor : 파라미터 네이밍 변경

iojkgh88 committed:
fix : null 변환값 변경

이름이 들어있지 않은것이 명시적으로 비공개임을 의미하지 않으므로
빈 형태로 처리.

iojkgh88 committed:
refactor : verifySeparated while문으로 변경

index를 사용하는 for문은 자연스러워보이지 않음.
reduce를 사용해보고자 하였으나 ExceptionInInitializerError발생으로 while문 사용

iojkgh88 committed:
fix : 오기입 수정

iojkgh88 committed:
merge : 기존 commit사항 merge

iojkgh88 committed:
feat : 사용가능시간반환메서드 구현

iojkgh88 committed:
feat : 사용가능시간반환메서드 구현

iojkgh88 committed:
feat : project에 MeetingTimeUnit추가

iojkgh88 committed:
feat : 가능시간뽑는 메서드 추가

iojkgh88 committed:
feat : 가능시간뽑는 메서드 추가

iojkgh88 committed:
Merge branch 'feature/timePlan' into feature/meeting

iojkgh88 committed:
refactor : TimePlan과 Event양방향 연관관계로 변경

Scheduling을 위해 TimePlan외에 Event리스트를 따로 요청함을 줄이고자 변경

iojkgh88 committed:
feet : 미팅 스케줄링 구현중

hongseungjae committed:
test: Project Mock Test 코드 작성

hongseungjae committed:
test: Test DisplayName 수정

iojkgh88 committed:
fix : 프로젝트에 MinuteUnit 제거

iojkgh88 committed:
refactor : 예외DTO ,예외메시지 및 ApplicationException 구조개선

1. 예외 DTO에 존재하는 상태코드, 에러코드 삭제
2. 예외메시지를 messages.properties에서 관리하도록 변경
3. ApplicationException 생성자 파라미터값을 없애기 위해
   에러코드 및 반환할 상태값을 미리 담는 형태로 변경

iojkgh88 committed:
feat : properties utf-8로 인코딩

iojkgh88 committed:
feat : properties utf-8로 인코딩

hongseungjae committed:
feat: invite service 추가

Commit:
refactor: 예외처리 개선

hongseungjae committed:
Merge branch 'feature/projectException' of https://github.com/f-lab-edu/quick-together into feature/project4

 Conflicts:
	src/main/java/com/flab/quicktogether/participant/application/ParticipantService.java
	src/test/java/com/flab/quicktogether/project/domain/ProjectTest.java

hongseungjae committed:
refactor: Exception클래스 변수에 final 활용하여 수정

hongseungjae committed:
refactor: JpaRepository에 의해 자동 추가되는 부분 제거

hongseungjae committed:
feat: invite serivce 기능 추가, Mock test코드 작성

hongseungjae committed:
fix: Application handler 로깅 수정

hongseungjae committed:
fix: addMember -> addParticipant 네이밍 수정

hongseungjae committed:
fix: addMember -> addParticipant 네이밍 수정

hongseungjae committed:
fix: skillStack -> position 네이밍 수정

hongseungjae committed:
feat: Invite Controller 추가

hongseungjae committed:
refactor: findInvite 사용 전 프로젝트와 멤버를 검증하게 작성

hongseungjae committed:
fix: Participant에 데이터가 추가 되므로 post로 수정

hongseungjae committed:
feat: 프로젝트 입장 기능 추가

hongseungjae committed:
feat: 프로젝트 입장 기능 추가

hongseungjae committed:
refactor: Exception throw 부분 람다식으로 통일성 있게 작성

hongseungjae committed:
test: Repository관련 스터빙

f-lab-fi commented on src/main/java/com/flab/quicktogether/globalsetting/domain/exception/ExceptionResponse.java:
```diff
@@ -0,0 +1,34 @@
+        this.path = path;
+    }
+
+    public ExceptionResponse(String message, String path,List<ObjectError> allErrors) {
+        this.message = message;
+        this.path = path;
+        allErrors.stream()
+                .map(allError -> (FieldError) allError)
+                .forEach(fieldError -> errors.add(new ErrorInfo(fieldError.getField(), fieldError.getRejectedValue(), fieldError.getDefaultMessage())));
+    }
```
서로 다른 예외들도 일관성있게 처리되도록 잘 만들어보셨네요.

hongseungjae replied:
넵 프론트엔드쪽에서도 일관된 로그 형식을 받아야 처리하기 편하기 때문에 같은 형식을 반환하고 있습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/meeting/application/MeetingSchedularService.java:
```diff
@@ -0,0 +1,54 @@
+    private final TimePlanRepository timePlanRepository;
+    private final Scheduler scheduler;
+
+    public List<TimeBlock> suggestAvailableTime(Long loginMemberId, Long projectId, AvailableTimeRequestDto availableTimeRequestDto) {
+        //MinuteUnit meetingTimeUnit = findProject(projectId).getMeetingTimeUnit();
+        MinuteUnit meetingTimeUnit = null;
+        LocalDate roughlyStartDate = availableTimeRequestDto.getRoughlyStartDate();
```
뭔가 문제가 있어서 null로 처리해두신건가요~?

f-lab-fi commented on src/main/java/com/flab/quicktogether/project/presentation/dto/request/InviteMemberRequest.java:
```diff
@@ -0,0 +1,18 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class InviteMemberRequest {
+    @NotNull
+    private Long requestMemberId;
+
+    @NotNull
+    private Long invitedMemberId;
+}
```
requestMemberId는 사용자가 Body로 보낼게 아니라 인증 수단을 통해 가져와야하는 정보 아닐까요~?

f-lab-fi replied:
[이부분](https://github.com/f-lab-edu/quick-together/pull/3/files/19854cff21755b5318f056f6d0fa15867459efb9..5c0bab4224b818a83547411ef84d2cc35fa6b906#diff-5e6766865273d75949a93a5a22fce1e5b681324ab50427d82d782f7fdf906c55R31-R39)과 관련된 내용입니다.

hongseungjae replied:
아 넵 맞습니다.
수정하겠습니다.

hongseungjae committed:
refactor: Project - Participant OneToMany 양방향 생성

f-lab-fi commented on src/main/java/com/flab/quicktogether/project/application/ProjectInviteService.java:
```diff
@@ -0,0 +1,103 @@
+        return invite;
+    }
+
+    @Transactional
+    public void inviteMember(Long projectId, Long requestMemberId, Long invitedMemberId) {
+        Member requestMember = findMember(requestMemberId);
+        Member invitedMember = findMember(invitedMemberId);
+        Project project = findProject(projectId);
+        checkProjectParticipation(projectId, invitedMemberId);
+
+        Participant requestParticipant = findParticipant(projectId, requestMemberId);
+        requestParticipant.checkPermission();
+
+        isInvited(projectId, invitedMemberId);
+        inviteRepository.save(Invite.inviteMember(project, requestMember, invitedMember));
+    }
```
findMember 메서드는 조회시 인덱스를 탈거기 때문에 비용이 크진 않겠지만, 생각해볼만한 내용이 있습니다.
requestMember는 처음에 조회되고 있지만, 실제 사용되는 것은 마지막에 있는 Invite.inviteMember 메서드 부분입니다.
그럼 결국 미리 조회를 해왔는데 도중에 예외가 발생하게되는 경우 requestMember를 조회했던건 아무 의미가 없어지는겁니다.
만약 이 연산이 무거운 연산이었다면 매우 비효율적인 코드가 됐을겁니다. 이런 문제를 피하려면 requestMember가 필요해지는 시점 직전에 조회를 해오시는게 적절해보입니다.

hongseungjae replied:
아 넵
지금 보니 member를 얻어오고 전혀 사용하지 않다가 마지막에 save될 때만 사용하고 있습니다.
파이님 말씀처럼 비효율적인 코드가 될 수 있을 것 같습니다.
수정하겠습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/project/domain/EnterStatus.java:
```diff
@@ -0,0 +1,5 @@
+package com.flab.quicktogether.project.domain;
+
+public enum EnterStatus {
+    ACCEPT, REJECT, WAIT
+}
```
EnterStatus와 InviteStatus가 현재는 같은데, 혹시 나중에 다른 상태들을 가질 가능성이 높을까요~?

hongseungjae replied:
나중에 다른 상태를 가질 가능성이 낮아보여서, 한 개로 합쳐도 무방할 것 같습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/project/presentation/controller/ProjectEnterController.java:
```diff
@@ -0,0 +1,69 @@
+        return ResponseEntity.ok(new ProjectEnterResponse(enter));
+    }
+
+
+    /**
+     * 프로젝트에 입장 신청
+     */
+    @PostMapping("/projects/{projectId}/enters")
+    public ResponseEntity enterMember(@PathVariable("projectId") Long projectId,
+                                      @RequestBody @Valid EnterMemberRequest enterMemberRequest) {
+
+        projectEnterService.requestEnterProject(projectId, enterMemberRequest.getEnterMemberId());
+        URI location = URI.create(String.format("/projects/%d/members/%d/enters", projectId, enterMemberRequest.getEnterMemberId()));
+        return ResponseEntity.created(location).build();
+    }
+
+    /**
+     * 프로젝트 입장 신청 수락
+     */
+    @PostMapping("/projects/{projectId}/members/{adminMemberId}/enters")
+    public ResponseEntity acceptEnter(@PathVariable("projectId") Long projectId,
+                                      @Login @PathVariable("adminMemberId") Long adminMemberId,
+                                      @RequestBody @Valid EnterMemberRequest enterMemberRequest) {
+
+        projectEnterService.acceptEnter(projectId, adminMemberId,enterMemberRequest.getEnterMemberId());
+        return ResponseEntity.ok().build();
+    }
+
+    /**
+     * 프로젝트 입장 신청 거절
+     */
+    @DeleteMapping("/projects/{projectId}/members/{adminMemberId}/enters")
+    public ResponseEntity rejectEnter(@PathVariable("projectId") Long projectId,
+                                      @Login @PathVariable("adminMemberId") Long adminMemberId,
+                                      @RequestBody @Valid EnterMemberRequest enterMemberRequest) {
+
+        projectEnterService.rejectEnter(projectId, adminMemberId,enterMemberRequest.getEnterMemberId());
+        return ResponseEntity.ok().build();
+    }
+}
```
여기 있는 각각의 경로에 있는 {enterMemberId}와 {adminMemberId} 들은 Path로 받는게 정상적인걸까요?
사용자가 보내주는 값을 사용하는게 아니라 사용자의 인증 정보에서 가져와야 하는건 아닐지 궁금합니다.

hongseungjae replied:
아 넵 맞습니다.
로그인 부분 좀 더 다듬으면서 수정해보겠습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/project/presentation/controller/ProjectInviteController.java:
```diff
@@ -0,0 +1,62 @@
+
+    /**
+     * 프로젝트에 초대된 멤버 조회
+     */
+    @GetMapping("/projects/{projectId}/members/{invitedMemberId}/invites")
+    public ResponseEntity inviteMember(@PathVariable("projectId") Long projectId, @PathVariable("invitedMemberId") Long invitedMemberId) {
+
```
{invitedMemberId}도 마찬가지입니다. :)

https://github.com/f-lab-edu/quick-together/pull/3/files/19854cff21755b5318f056f6d0fa15867459efb9..5c0bab4224b818a83547411ef84d2cc35fa6b906#r1068250855

hongseungjae replied:
넵 수정하겠습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/timeplan/domain/TimePlan.java:
```diff
@@ -34,31 +34,60 @@ public class TimePlan {
         Collections.sort(ableRoutines);
-        for (int i = 0; i < ableRoutines.size()-1; i++) {
-            RegularTimeBlock cur = ableRoutines.get(i).getRegularTimeBlock();
-            RegularTimeBlock next = ableRoutines.get(i + 1).getRegularTimeBlock();
 
-            if (!cur.isSeperatedFrom(next)) {
+        Iterator<AbleRoutine> iterator = ableRoutines.iterator();
+        RegularTimeBlock curRtb = iterator.next().getRegularTimeBlock();
+
+        while (iterator.hasNext()) {
+            RegularTimeBlock nextRtb = iterator.next().getRegularTimeBlock();
+
+            if (!curRtb.isSeperatedFrom(nextRtb)) {
                 throw new AbleRoutineOverlapException();
             }
+            curRtb = nextRtb;
         }
```
기존 코드보다 가독성 있게 바뀐 것 같네요. :)

f-lab-fi commented on src/test/java/com/flab/quicktogether/project/application/MockProjectInviteServiceTest.java:
```diff
@@ -0,0 +1,136 @@
+    public void inviteMemberException(){
+
+        //given
+        Project project = spy(Project.class);
+        Member requestMember = spy(Member.class);
+        Member invitedMember = spy(Member.class);
+        Participant participant = spy(Participant.class);
+        participant.changeParticipantRole(ParticipantRole.ROLE_ADMIN);
+        Invite invite = mock(Invite.class);
+        long requestMemberId = 1;
+        long invitedMemberId = 2;
```
spy랑 mock을 구분해서 잘 모킹하셨네요. 👍

hongseungjae committed:
feat: 초대, 참여 이벤트 추가

hongseungjae committed:
feat: 세션 로그인 구현

hongseungjae committed:
feat: Jwt 로그인 구현, 회원가입 구현

hongseungjae committed:
Merge branch 'feature/project4_Login' into feature/project4_merge_event_login

hongseungjae committed:
test: oneToMany로 인해 테스트 코드 일부 수정

hongseungjae committed:
fix: 초대, 입장 기능 권한 수정

hongseungjae committed:
fix: 구성원 전체 조회 전, 프로젝트 find 추가

hongseungjae committed:
fix: member 객체 사용 될 때 find 하기

hongseungjae committed:
refactor: enterStatus와 inviteStatus를 ProjectJoinStatus로 합침

hongseungjae committed:
test: invite mock test 실패 케이스 수정

f-lab-fi commented on src/main/java/com/flab/quicktogether/common/auth/config/jwt/TokenDto.java:
```diff
@@ -0,0 +1,18 @@
+import io.jsonwebtoken.Claims;
+import lombok.Getter;
+
+@Getter
+public class TokenDto {
+
+    private Claims token;
+    private String memberName;
+    private Long memberId;
+
+    public TokenDto(Claims token, String memberName, Object memberId) {
+        this.token = token;
+        this.memberName = memberName;
+        this.memberId = Long.valueOf(String.valueOf(memberId));
+    }
+}
```
멘토링 시간에 이야기했던 것처럼 JWT는 토큰 자체적으로 어떤 정보를 저장할 수 있다 특징이 있기 때문에 자주 사용되는 정보가 있다면 지금 하신것처럼 토큰 내에 저장하는게 좋은 방법이 될 수가 있습니다.
DB에 조회하지 않아도 되니까요.

f-lab-fi commented on src/main/java/com/flab/quicktogether/member/application/MemberService.java:
```diff
@@ -1,28 +1,78 @@
+
+    }
+
+    private void validateDuplicateMemberName(MemberRequest memberRequest) {
+        memberRepository.findByMemberName(memberRequest.getMemberName()).ifPresent(member -> {
+                    throw new DuplicateMemberNameException();
+                });
+    }
```
```java
        memberRepository.findByMemberName(memberRequest.getMemberName())
                          .ifPresent(member -> {
                              throw new DuplicateMemberNameException();
                          });
```

메서드 체이닝을 하는 상황에서는 가급적 메서드간 개행을 시켜주세요.
이쪽이 더 가독성이 좋아보이지 않나요? :)

hongseungjae replied:
넵 맞습니다.
기존 작성된 코드는 메서드 체이닝이 길어질수록 더 가독성이 안 좋아질 것 같습니다.
수정하겠습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/member/presentation/LoginController.java:
```diff
@@ -0,0 +1,46 @@
+        return ResponseEntity.ok(new MemberIdResponse(loginId));
+    }
+
+    @RequestMapping(path = "/login", method = RequestMethod.GET)
+    public ResponseEntity isLogin(@Login Long memberId) {
+        return ResponseEntity.ok(new MemberIdResponse(memberId));
+    }
```
이건 로그인된 상태인지 체크하는 API인가요?
어떤 상황에서 호출되는걸까요~?

hongseungjae replied:
넵
post 방식 로그인 이후
get 방식으로 간단히 테스트할 용도로 사용되었습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/participant/domain/Participants.java:
```diff
@@ -0,0 +1,56 @@
+    public Participants() {
+    }
+
+    public void isParticipantNot(Long invitedMemberId) {
+        participants.stream()
+                .filter(participant -> participant.getMember().getId().equals(invitedMemberId))
+                .findFirst()
+                .ifPresent(joinedParticipant -> {
+                    throw new DuplicateProjectParticipationException();
+                });
+    }
+
+    public void isAdmin(Long requestMemberId) {
+        participants.stream()
+                .filter(participant -> participant.getMember().getId().equals(requestMemberId))
+                .findFirst()
+                .ifPresentOrElse(Participant::checkPermission,
+                        () -> {
+                            throw new ParticipantNotFoundException();
+                        });
+    }
```
isXXX 는 일반적으로 Boolean 타입 리턴을 가지는 메서드에 붙는 네이밍입니다.
이 상황에서는 checkXXX가 더 적절한 이름이라고 생각됩니다. :)
메서드 네이밍 관련해서 한번 찾아보시고 바꿔보세요.

hongseungjae replied:
넵 
is/has/can 네이밍 같은 경우 보통 Boolean 타입 리턴을 갖는 메서드라고 합니다.
check나 validate로 수정하겠습니다.

f-lab-fi commented on src/main/java/com/flab/quicktogether/project/event/Events.java:
```diff
@@ -0,0 +1,17 @@
+package com.flab.quicktogether.project.event;
+
+import org.springframework.context.ApplicationEventPublisher;
+
+public class Events {
+    private static ApplicationEventPublisher publisher;
+
+    static void setPublisher(ApplicationEventPublisher publisher) {
+        Events.publisher = publisher;
+    }
+
+    public static void raise(Object event) {
+        if (publisher != null) {
+            publisher.publishEvent(event);
+        }
+    }
+}
```
💯

hongseungjae committed:
test: 프로젝트 생성 관련 잘못된 테스트 수정

hongseungjae committed:
refactor: 프로젝트 패키지내 부가 서비스들 따로 패키지화

hongseungjae committed:
feat: Participant Auth Exception 추가

hongseungjae committed:
feat: Participant Auth Exception 추가

hongseungjae committed:
feat: Participant Auth Exception 추가

hongseungjae committed:
feat: Project에 post 기능 추가

hongseungjae committed:
feat: Project에 입장 조건 추가

hongseungjae committed:
fix: NotAuthorizedParticipantException 추가로인한 테스트 코드 수정

hongseungjae committed:
refactor: 반환값 없는 메서드 네이밍 is -> check로 수정

hongseungjae committed:
fix: project엔티티와 participants엔티티 orphanRemoval = true 추가

hongseungjae committed:
refactor: String 클래스 isEmpty 사용

hongseungjae committed:
refactor: Project 단순조회, 상세조회 join 이용

hongseungjae committed:
refactor: Post 도메인 pk이름 id -> post_id 변경

f-lab-fi commented on src/main/java/com/flab/quicktogether/project/domain/Project.java:
```diff
@@ -117,8 +116,12 @@ public void changeProjectDescriptionInfo(ProjectDescriptionInfo editProjectDescr
-    public void settingLikes(Long likes) {
-        this.likes = likes;
+    public void checkJoinProject(){
+        if(projectStatus.equals(ProjectStatus.OPEN)){
+            //do nothing
+        }else {
+           throw new JoinProjectException();
+        }
```
이 분기문 ProjectStatus를 활용해서 없애볼 수 있지 않을까요~?
